### PR TITLE
Include comments with template argument names in Cpp code from EmitC

### PR DESCRIPTION
**Pull request by lmendesp-amd**

### CODE REVIEW

File: mlir/lib/Target/Cpp/TranslateToCpp.cpp [hunk 1]
```diff
@@ -659,11 +659,31 @@ static LogicalResult printOperation(CppEmitter &emitter,
...
+
   if (callOpaqueOp.getTemplateArgs()) {
     os << "<";
-    if (failed(interleaveCommaWithError(*callOpaqueOp.getTemplateArgs(), os,
-                                        emitArgs)))
-      return failure();
+    if (callOpaqueOp.getTemplateArgNames() &&
+        !callOpaqueOp.getTemplateArgNames()->empty()) {
+      if (failed(interleaveCommaWithError(
+              llvm::zip(*callOpaqueOp.getTemplateArgs(),
```
Comment (cferry-AMD) [inline mlir/lib/Target/Cpp/TranslateToCpp.cpp:677]:
```suggestion
              llvm::zip_equal(*callOpaqueOp.getTemplateArgs(),
```

**Review by cferry-AMD — APPROVED** [review]
LGTM, thanks!

File: mlir/test/Target/Cpp/template_arg_names.mlir [hunk 2]
```diff
@@ -0,0 +1,20 @@
+// RUN: mlir-translate -mlir-to-cpp %s | FileCheck %s -check-prefix=CPP-DEFAULT
+
+// CPP-DEFAULT: void basic(int32_t v1, float v2) {
```
Comment (mgehre-amd) [inline mlir/test/Target/Cpp/template_arg_names.mlir:3]:
```suggestion
// CPP-DEFAULT-LABEL: void basic
```

File: mlir/test/Target/Cpp/template_arg_names.mlir [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+  emitc.call_opaque "kernel3"(%arg0, %arg1)  : (i32, f32) -> ()
+// CPP-DEFAULT:   kernel3(v1, v2);
+  emitc.call_opaque "kernel4"(%arg0, %arg1)  {template_arg_names = ["N", "P"], template_args = [42 : i32, 56]} : (i32, f32) -> ()
+// CPP-DEFAULT:   kernel4</*N=*/42, /*P=*/56>(v1, v2);
+  emitc.call_opaque "kernel4"(%arg0, %arg1)  {template_arg_names = ["N"], template_args = [#emitc.opaque<"42">]} : (i32, f32) -> ()
+// CPP-DEFAULT:   kernel4</*N=*/42>(v1, v2);
+  return
+// CPP-DEFAULT:   return;
+}
+// CPP-DEFAULT: }
```
Comment (mgehre-amd) [inline mlir/test/Target/Cpp/template_arg_names.mlir:18]: no need to check those here
```suggestion
}
```

File: mlir/test/Target/Cpp/template_arg_names.mlir [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+  emitc.call_opaque "kernel1"()  : () -> ()
+// CPP-DEFAULT:   kernel1();
+  %0 = emitc.call_opaque "kernel2"(%arg0)  : (i32) -> i16
+// CPP-DEFAULT:   int16_t v3 = kernel2(v1);
+  emitc.call_opaque "kernel3"(%arg0, %arg1)  : (i32, f32) -> ()
+// CPP-DEFAULT:   kernel3(v1, v2);
+  emitc.call_opaque "kernel4"(%arg0, %arg1)  {template_arg_names = ["N", "P"], template_args = [42 : i32, 56]} : (i32, f32) -> ()
+// CPP-DEFAULT:   kernel4</*N=*/42, /*P=*/56>(v1, v2);
+  emitc.call_opaque "kernel4"(%arg0, %arg1)  {template_arg_names = ["N"], template_args = [#emitc.opaque<"42">]} : (i32, f32) -> ()
+// CPP-DEFAULT:   kernel4</*N=*/42>(v1, v2);
```
Comment (mgehre-amd) [inline mlir/test/Target/Cpp/template_arg_names.mlir:14]: better not depend on the variable names (here and above)
```suggestion
// CPP-DEFAULT:   kernel4</*N=*/42>(
```

File: mlir/lib/Dialect/EmitC/IR/EmitC.cpp [hunk 5]
```diff
@@ -353,6 +353,14 @@ LogicalResult emitc::CallOpaqueOp::verify() {
...
     }
+
+    if (std::optional<ArrayAttr> templateArgNames = getTemplateArgNames()) {
+      if ((*templateArgNames).size() &&
+          (*templateArgNames).size() != (*templateArgsAttr).size()) {
+        return emitOpError("number of template argument names must be equal to "
+                           "number of template arguments");
+      }
+    }
   }
```
Comment (mgehre-amd) [inline mlir/lib/Dialect/EmitC/IR/EmitC.cpp:364]: Can we please verify that we have no template arg names whenever we have no template args, and add a test for that?

**Comment by mgehre-amd** [comment]
Thank you! I added a small comments, but looks good overall.

File: mlir/lib/Target/Cpp/TranslateToCpp.cpp [hunk 6]
```diff
@@ -659,11 +659,31 @@ static LogicalResult printOperation(CppEmitter &emitter,
...
+    return emitArgs(attr);
+  };
+
   if (callOpaqueOp.getTemplateArgs()) {
     os << "<";
-    if (failed(interleaveCommaWithError(*callOpaqueOp.getTemplateArgs(), os,
-                                        emitArgs)))
-      return failure();
+    if (callOpaqueOp.getTemplateArgNames() &&
+        !callOpaqueOp.getTemplateArgNames()->empty()) {
```
[the remark this answers is not in the document]
Comment (lmendesp-amd) [reply mlir/lib/Target/Cpp/TranslateToCpp.cpp:675]: We made the default to be passing an empty `AttrArray` to the `build` method.

File: mlir/lib/Dialect/EmitC/IR/EmitC.cpp (hunk 5 above)
Comment (lmendesp-amd) [reply mlir/lib/Dialect/EmitC/IR/EmitC.cpp:369]: Added.

File: mlir/lib/Target/Cpp/TranslateToCpp.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (mgehre-amd) [reply mlir/lib/Target/Cpp/TranslateToCpp.cpp:675]: If `getTemplateArgNames` is empty, we already know that `getTemplateArgs` is also empty, right?
So in that case, `interleaveCommaWithError` would be correct (and do nothing), right?

**Review by mgehre-amd — APPROVED** [review]

**Pull request merged**